File: trabalho-final-sistemas-distribuidos/go-engine/internal/sparql/parser.go
// Package sparql parses the SPARQL SELECT subset exercised by FedShop.
package sparql

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"unicode"
)

// TermKind identifies a SPARQL term.
type TermKind uint8

const (
	TermVariable TermKind = iota
	TermIRI
	TermLiteral
	TermBare
)

// Term is one RDF term in a triple pattern.
type Term struct {
	Kind  TermKind
	Value string
}

// TriplePattern is a basic graph pattern statement.
type TriplePattern struct {
	ID                         int
	Subject, Predicate, Object Term
}

// Key returns the FedShop composition-compatible triple representation.
func (t TriplePattern) Key() string {
	return strings.Join([]string{t.Subject.SPARQL(), t.Predicate.SPARQL(), t.Object.SPARQL()}, " ")
}

// Variables returns the unique variables in statement order.
func (t TriplePattern) Variables() []string {
	seen := map[string]bool{}
	var result []string
	for _, term := range []Term{t.Subject, t.Predicate, t.Object} {
		if term.Kind == TermVariable && !seen[term.Value] {
			seen[term.Value] = true
			result = append(result, term.Value)
		}
	}
	return result
}

// SPARQL renders a term without losing its RDF lexical representation.
func (t Term) SPARQL() string {
	switch t.Kind {
	case TermVariable:
		return "?" + t.Value
	case TermIRI:
		return "<" + t.Value + ">"
	default:
		return t.Value
	}
}

// Group is a graph-pattern group.
type Group struct {
	Triples   []TriplePattern
	Filters   []string
	Optionals []*Group
	Unions    [][2]*Group
}

// OrderCondition is one ORDER BY expression.
type OrderCondition struct {
	Expression string
	Ascending  bool
}

// Query is the FedShop-supported SELECT algebra.
type Query struct {
	Prefixes map[string]string
	Select   []string
	Distinct bool
	Where    *Group
	OrderBy  []OrderCondition
	Offset   int
	Limit    int
}

// Triples returns every triple in stable source order.
func (q Query) Triples() []TriplePattern {
	var result []TriplePattern
	var visit func(*Group)
	visit = func(group *Group) {
		result = append(result, group.Triples...)
		for _, pair := range group.Unions {
			visit(pair[0])
			visit(pair[1])
		}
		for _, optional := range group.Optionals {
			visit(optional)
		}
	}
	visit(q.Where)
	return result
}

var (
	prefixPattern   = regexp.MustCompile(`(?im)^\s*PREFIX\s+([A-Za-z][\w-]*):\s*<([^>]+)>`)
	selectPattern   = regexp.MustCompile(`(?is)\bSELECT\s+(DISTINCT\s+)?(.+?)\s+(?:WHERE\s*)?\{`)
	variablePattern = regexp.MustCompile(`\?([A-Za-z_][\w-]*)`)
	limitPattern    = regexp.MustCompile(`(?is)\bLIMIT\s+(\d+)`)
	offsetPattern   = regexp.MustCompile(`(?is)\bOFFSET\s+(\d+)`)
	orderPattern    = regexp.MustCompile(`(?is)\bORDER\s+BY\s+(.+?)(?:\bLIMIT\b|\bOFFSET\b|$)`)
)

// Parse parses the SELECT subset used by all FedShop query templates.
func Parse(input string) (Query, error) {
	clean := stripComments(input)
	if regexp.MustCompile(`(?i)\b(GRAPH|SERVICE|BIND|VALUES|MINUS|GROUP\s+BY|HAVING)\b`).MatchString(clean) {
		return Query{}, fmt.Errorf("parse query: unsupported graph-pattern feature")
	}
	prefixes := map[string]string{}
	for _, match := range prefixPattern.FindAllStringSubmatch(clean, -1) {
		prefixes[match[1]] = match[2]
	}
	selectMatch := selectPattern.FindStringSubmatchIndex(clean)
	if selectMatch == nil {
		return Query{}, fmt.Errorf("parse query: expected SELECT and graph pattern")
	}
	selectMatchText := selectPattern.FindStringSubmatch(clean)
	selected := unique(variablePattern.FindAllStringSubmatch(selectMatchText[2], -1))
	if strings.TrimSpace(selectMatchText[2]) == "*" {
		selected = nil
	}
	open := strings.Index(clean[selectMatch[0]:selectMatch[1]], "{") + selectMatch[0]
	close, err := matching(clean, open, '{', '}')
	if err != nil {
		return Query{}, err
	}
	group, err := parseGroup(clean[open+1:close], prefixes)
	if err != nil {
		return Query{}, err
	}
	assignTripleIDs(group)
	query := Query{Prefixes: prefixes, Select: selected, Distinct: strings.TrimSpace(selectMatchText[1]) != "", Where: group, Limit: -1}
	if match := limitPattern.FindStringSubmatch(clean[close+1:]); match != nil {
		query.Limit, err = strconv.Atoi(match[1])
		if err != nil {
			return Query{}, fmt.Errorf("parse LIMIT: %w", err)
		}
	}
	if match := offsetPattern.FindStringSubmatch(clean[close+1:]); match != nil {
		query.Offset, err = strconv.Atoi(match[1])
		if err != nil {
			return Query{}, fmt.Errorf("parse OFFSET: %w", err)
		}
	}
	if match := orderPattern.FindStringSubmatch(clean[close+1:]); match != nil {
		query.OrderBy = parseOrder(match[1])
	}
	return query, nil
}

func unique(matches [][]string) []string {
	seen := map[string]bool{}
	var result []string
	for _, match := range matches {
		if !seen[match[1]] {
			seen[match[1]] = true
			result = append(result, match[1])
		}
	}
	return result
}

func parseOrder(input string) []OrderCondition {
	var result []OrderCondition
	for len(strings.TrimSpace(input)) > 0 {
		input = strings.TrimSpace(input)
		ascending := true
		expr := ""
		upper := strings.ToUpper(input)
		if strings.HasPrefix(upper, "DESC(") || strings.HasPrefix(upper, "ASC(") {
			ascending = strings.HasPrefix(upper, "ASC(")
			open := strings.Index(input, "(")
			close, err := matching(input, open, '(', ')')
			if err != nil {
				return result
			}
			expr, input = strings.TrimSpace(input[open+1:close]), input[close+1:]
		} else if strings.HasPrefix(input, "?") {
			end := 1
			for end < len(input) && (unicode.IsLetter(rune(input[end])) || unicode.IsDigit(rune(input[end])) || input[end] == '_') {
				end++
			}
			expr, input = input[:end], input[end:]
		} else {
			// FedShop q10 uses xsd:double(str(?price)); retain it as one expression.
			end := len(input)
			expr, input = strings.TrimSpace(input[:end]), ""
		}
		result = append(result, OrderCondition{Expression: expr, Ascending: ascending})
	}
	return result
}

func assignTripleIDs(group *Group) {
	next := 0
	var visit func(*Group)
	visit = func(g *Group) {
		for i := range g.Triples {
			g.Triples[i].ID = next
			next++
		}
		for _, union := range g.Unions {
			visit(union[0])
			visit(union[1])
		}
		for _, optional := range g.Optionals {
			visit(optional)
		}
	}
	visit(group)
}

func parseGroup(body string, prefixes map[string]string) (*Group, error) {
	group := &Group{}
	for pos := 0; pos < len(body); {
		pos = skipSpace(body, pos)
		if pos >= len(body) {
			break
		}
		if hasKeyword(body, pos, "OPTIONAL") {
			pos = skipSpace(body, pos+len("OPTIONAL"))
			if pos >= len(body) || body[pos] != '{' {
				return nil, fmt.Errorf("parse OPTIONAL: expected {")
			}
			close, err := matching(body, pos, '{', '}')
			if err != nil {
				return nil, err
			}
			child, err := parseGroup(body[pos+1:close], prefixes)
			if err != nil {
				return nil, err
			}
			group.Optionals = append(group.Optionals, child)
			pos = close + 1
			continue
		}
		if hasKeyword(body, pos, "FILTER") {
			pos = skipSpace(body, pos+len("FILTER"))
			if pos >= len(body) {
				return nil, fmt.Errorf("parse FILTER: expected expression")
			}
			exprStart, open := pos, pos
			if body[open] != '(' {
				open = strings.IndexByte(body[pos:], '(')
				if open < 0 {
					return nil, fmt.Errorf("parse FILTER: expected expression")
				}
				open += pos
			}
			close, err := matching(body, open, '(', ')')
			if err != nil {
				return nil, err
			}
			if open == pos {
				group.Filters = append(group.Filters, strings.TrimSpace(body[pos+1:close]))
			} else {
				group.Filters = append(group.Filters, strings.TrimSpace(body[exprStart:close+1]))
			}
			pos = close + 1
			continue
		}
		if body[pos] == '{' {
			close, err := matching(body, pos, '{', '}')
			if err != nil {
				return nil, err
			}
			left, err := parseGroup(body[pos+1:close], prefixes)
			if err != nil {
				return nil, err
			}
			next := skipSpace(body, close+1)
			if !hasKeyword(body, next, "UNION") {
				return nil, fmt.Errorf("parse subgroup: only UNION groups are supported")
			}
			next = skipSpace(body, next+len("UNION"))
			if next >= len(body) || body[next] != '{' {
				return nil, fmt.Errorf("parse UNION: expected second group")
			}
			rightClose, err := matching(body, next, '{', '}')
			if err != nil {
				return nil, err
			}
			right, err := parseGroup(body[next+1:rightClose], prefixes)
			if err != nil {
				return nil, err
			}
			group.Unions = append(group.Unions, [2]*Group{left, right})
			pos = rightClose + 1
			continue
		}
		statementEnd := findStatementEnd(body, pos)
		triples, err := parseStatement(body[pos:statementEnd], prefixes)
		if err != nil {
			return nil, err
		}
		group.Triples = append(group.Triples, triples...)
		pos = statementEnd
		if pos < len(body) && body[pos] == '.' {
			pos++
		}
	}
	return group, nil
}

func parseStatement(input string, prefixes map[string]string) ([]TriplePattern, error) {
	tokens := tokenize(strings.TrimSpace(input))
	if len(tokens) < 3 {
		return nil, fmt.Errorf("parse triple near %q: expected subject predicate object", strings.TrimSpace(input))
	}
	subject, err := parseTerm(tokens[0], prefixes)
	if err != nil {
		return nil, err
	}
	var result []TriplePattern
	var predicate Term
	for i := 1; i < len(tokens); {
		if tokens[i] == ";" {
			i++
			continue
		}
		if tokens[i] == "," {
			i++
			if i >= len(tokens) {
				return nil, fmt.Errorf("parse comma object")
			}
			object, termErr := parseTerm(tokens[i], prefixes)
			if termErr != nil {
				return nil, termErr
			}
			result = append(result, TriplePattern{Subject: subject, Predicate: predicate, Object: object})
			i++
			continue
		}
		if i+1 >= len(tokens) {
			return nil, fmt.Errorf("parse triple: missing object")
		}
		predicate, err = parseTerm(tokens[i], prefixes)
		if err != nil {
			return nil, err
		}
		object, termErr := parseTerm(tokens[i+1], prefixes)
		if termErr != nil {
			return nil, termErr
		}
		result = append(result, TriplePattern{Subject: subject, Predicate: predicate, Object: object})
		i += 2
	}
	return result, nil
}

func parseTerm(token string, prefixes map[string]string) (Term, error) {
	if strings.HasPrefix(token, "?") {
		return Term{Kind: TermVariable, Value: token[1:]}, nil
	}
	if strings.HasPrefix(token, "<") && strings.HasSuffix(token, ">") {
		return Term{Kind: TermIRI, Value: token[1 : len(token)-1]}, nil
	}
	if token == "a" {
		return Term{Kind: TermIRI, Value: "http://www.w3.org/1999/02/22-rdf-syntax-ns#type"}, nil
	}
	if strings.HasPrefix(token, `"`) {
		return Term{Kind: TermLiteral, Value: token}, nil
	}
	if colon := strings.IndexByte(token, ':'); colon >= 0 {
		if base, ok := prefixes[token[:colon]]; ok {
			return Term{Kind: TermIRI, Value: base + token[colon+1:]}, nil
		}
	}
	return Term{Kind: TermBare, Value: token}, nil
}

func tokenize(input string) []string {
	var tokens []string
	for i := 0; i < len(input); {
		if unicode.IsSpace(rune(input[i])) {
			i++
			continue
		}
		start := i
		switch input[i] {
		case '<':
			i++
			for i < len(input) && input[i] != '>' {
				i++
			}
			if i < len(input) {
				i++
			}
		case '"':
			i++
			escaped := false
			for i < len(input) {
				if !escaped && input[i] == '"' {
					i++
					break
				}
				escaped = !escaped && input[i] == '\\'
				if input[i] != '\\' {
					escaped = false
				}
				i++
			}
			for i < len(input) && !unicode.IsSpace(rune(input[i])) && input[i] != ';' && input[i] != ',' {
				i++
			}
		case ';', ',':
			i++
		default:
			for i < len(input) && !unicode.IsSpace(rune(input[i])) && input[i] != ';' && input[i] != ',' {
				i++
			}
		}
		tokens = append(tokens, input[start:i])
	}
	return tokens
}

func findStatementEnd(input string, start int) int {
	inIRI, inString, escaped := false, false, false
	for i := start; i < len(input); i++ {
		ch := input[i]
		if escaped {
			escaped = false
			continue
		}
		if inString && ch == '\\' {
			escaped = true
			continue
		}
		if !inString && ch == '<' && i+1 < len(input) && !unicode.IsSpace(rune(input[i+1])) && input[i+1] != '=' {
			inIRI = true
		} else if inIRI && ch == '>' {
			inIRI = false
		} else if !inIRI && ch == '"' {
			inString = !inString
		}
		if !inIRI && !inString && (ch == '.' || ch == '{' || ch == '}') {
			return i
		}
	}
	return len(input)
}

func matching(input string, open int, left, right byte) (int, error) {
	depth, inIRI, inString, escaped := 0, false, false, false
	for i := open; i < len(input); i++ {
		ch := input[i]
		if escaped {
			escaped = false
			continue
		}
		if inString && ch == '\\' {
			escaped = true
			continue
		}
		if !inString && ch == '<' && i+1 < len(input) && !unicode.IsSpace(rune(input[i+1])) && input[i+1] != '=' {
			inIRI = true
		} else if inIRI && ch == '>' {
			inIRI = false
		} else if !inIRI && ch == '"' {
			inString = !inString
		}
		if inIRI || inString {
			continue
		}
		if ch == left {
			depth++
		} else if ch == right {
			depth--
			if depth == 0 {
				return i, nil
			}
		}
	}
	return 0, fmt.Errorf("parse query: missing %q", right)
}

func stripComments(input string) string {
	var result strings.Builder
	for _, line := range strings.Split(input, "\n") {
		inIRI, inString, escaped := false, false, false
		for _, ch := range line {
			if escaped {
				result.WriteRune(ch)
				escaped = false
				continue
			}
			if inString && ch == '\\' {
				result.WriteRune(ch)
				escaped = true
				continue
			}
			if !inString && ch == '<' {
				inIRI = true
			} else if inIRI && ch == '>' {
				inIRI = false
			} else if !inIRI && ch == '"' {
				inString = !inString
			}
			if ch == '#' && !inIRI && !inString {
				break
			}
			result.WriteRune(ch)
		}
		result.WriteByte('\n')
	}
	return result.String()
}

func skipSpace(input string, pos int) int {
	for pos < len(input) && unicode.IsSpace(rune(input[pos])) {
		pos++
	}
	return pos
}
func hasKeyword(input string, pos int, keyword string) bool {
	if pos < 0 || pos+len(keyword) > len(input) || !strings.EqualFold(input[pos:pos+len(keyword)], keyword) {
		return false
	}
	return pos+len(keyword) == len(input) || !(unicode.IsLetter(rune(input[pos+len(keyword)])) || unicode.IsDigit(rune(input[pos+len(keyword)])) || input[pos+len(keyword)] == '_')
}

File: trabalho-final-sistemas-distribuidos/pyfedx/pyfedx_engine/parser.py
"""SPARQL and federation-config parsers."""

from __future__ import annotations

import re
from typing import Dict, Iterable, List, Optional, Tuple

from .model import Endpoint, GroupPattern, Query, RDF_TYPE, SparqlError, Triple

PREFIX_RE = re.compile(r"(?im)^\s*PREFIX\s+([A-Za-z][\w-]*):\s*<([^>]+)>")
SELECT_RE = re.compile(r"(?is)\bSELECT\s+(DISTINCT\s+)?(.+?)\s+(?:\bWHERE\b|\{)")
LIMIT_RE = re.compile(r"(?is)\bLIMIT\s+(\d+)")
VAR_RE = re.compile(r"\?([A-Za-z_][\w-]*)")
ENDPOINT_RE = re.compile(r"(?is)<([^>]+)>\s+a\s+sd:Service\s*;.*?sd:endpoint\s+\"([^\"]+)\"")
ORDERBY_RE = re.compile(r"(?is)\bORDER\s+BY\b(.+?)(?:\bLIMIT\b|\bOFFSET\b|$)")

# ---------------------------------------------------------------------------
# String / character utilities
# ---------------------------------------------------------------------------


def strip_comments(text: str) -> str:
    lines = []
    for line in text.splitlines():
        in_iri = in_string = escaped = False
        keep = []
        for ch in line:
            if escaped:
                keep.append(ch)
                escaped = False
                continue
            if ch == "\\" and in_string:
                keep.append(ch)
                escaped = True
                continue
            if ch == "<" and not in_string:
                in_iri = True
            elif ch == ">" and in_iri and not in_string:
                in_iri = False
            elif ch == '"' and not in_iri:
                in_string = not in_string
            if ch == "#" and not in_iri and not in_string:
                break
            keep.append(ch)
        lines.append("".join(keep))
    return "\n".join(lines)


def matching_brace(text: str, open_pos: int) -> int:
    depth = in_iri = in_string = escaped = False
    depth = 0
    for pos in range(open_pos, len(text)):
        ch = text[pos]
        if escaped:
            escaped = False
            continue
        if ch == "\\" and in_string:
            escaped = True
            continue
        if ch == "<" and not in_string:
            in_iri = True
        elif ch == ">" and in_iri and not in_string:
            in_iri = False
        elif ch == '"' and not in_iri:
            in_string = not in_string
        elif not in_iri and not in_string:
            if ch == "{":
                depth += 1
            elif ch == "}":
                depth -= 1
                if depth == 0:
                    return pos
    raise SparqlError("missing closing brace")


def where_body(text: str) -> str:
    match = re.search(r"(?is)\bWHERE\b\s*{", text)
    if not match:
        # No WHERE keyword — try bare { ... }
        open_pos = text.find("{")
        if open_pos == -1:
            raise SparqlError("unsupported query: no WHERE or { block found")
        close_pos = matching_brace(text, open_pos)
        return text[open_pos + 1 : close_pos]
    open_pos = text.find("{", match.start())
    close_pos = matching_brace(text, open_pos)
    return text[open_pos + 1 : close_pos]


# ---------------------------------------------------------------------------
# Tokenizer (for BGP and filter expressions)
# ---------------------------------------------------------------------------


def tokenize(body: str) -> List[str]:
    """Tokenize a SPARQL WHERE body (not filter expressions)."""
    token_re = re.compile(
        r"""
        <[^>]*>
        |"(?:\\.|[^"\\])*"(?:@[A-Za-z-]+|\^\^<[^>]+>|\^\^[A-Za-z][\w-]*:[\w.-]+)?
        |\?[A-Za-z_][\w-]*
        |[A-Za-z][\w-]*:[\w.-]+
        |[{}();,.]
        |[^\s{}();,.]+
        """,
        re.X,
    )
    return token_re.findall(strip_comments(body))


def tokenize_filter(expr: str) -> List[str]:
    """Tokenize a SPARQL filter expression, correctly disambiguating < IRI > from < operator."""
    tokens: List[str] = []
    i, n = 0, len(expr)
    while i < n:
        if expr[i].isspace():
            i += 1
            continue
        # Two-char operators
        if i + 1 < n and expr[i : i + 2] in ("&&", "||", "!=", "<=", ">="):
            tokens.append(expr[i : i + 2])
            i += 2
            continue
        # IRI or < comparison operator
        if expr[i] == "<":
            j = i + 1
            # IRI if next char is not whitespace and not '='
            if j < n and not expr[j].isspace() and expr[j] != "=":
                k = expr.find(">", j)
                if k != -1 and "\n" not in expr[j:k]:
                    tokens.append(expr[i : k + 1])
                    i = k + 1
                    continue
            tokens.append("<")
            i += 1
            continue
        if expr[i] in ">!=":
            tokens.append(expr[i])
            i += 1
            continue
        # Typed/language literal
        if expr[i] == '"':
            j = i + 1
            while j < n:
                if expr[j] == "\\" and j + 1 < n:
                    j += 2
                    continue
                if expr[j] == '"':
                    j += 1
                    break
                j += 1
            if j < n and expr[j : j + 2] == "^^":
                j += 2
                if j < n and expr[j] == "<":
                    k = expr.find(">", j)
                    j = k + 1 if k != -1 else j
                else:
                    while j < n and expr[j] not in " \t\n\r(),":
                        j += 1
            elif j < n and expr[j] == "@":
                j += 1
                while j < n and (expr[j].isalpha() or expr[j] == "-"):
                    j += 1
            tokens.append(expr[i:j])
            i = j
            continue
        # Variable
        if expr[i] == "?":
            j = i + 1
            while j < n and (expr[j].isalnum() or expr[j] == "_"):
                j += 1
            tokens.append(expr[i:j])
            i = j
            continue
        # Parens, comma, arithmetic operators
        if expr[i] in "(),+-*":
            tokens.append(expr[i])
            i += 1
            continue
        # Identifier / keyword / number
        j = i
        while j < n and expr[j] not in " \t\n\r<>!=&|(),+-*":
            j += 1
        if j > i:
            tokens.append(expr[i:j])
            i = j
        else:
            i += 1
    return tokens


# ---------------------------------------------------------------------------
# BGP parser (triples only, no FILTER/OPTIONAL/UNION)
# ---------------------------------------------------------------------------


def expand_term(term: str, prefixes: Dict[str, str]) -> str:
    if term == "a":
        return RDF_TYPE
    if term.startswith("?") or term.startswith("<") or term.startswith('"'):
        return term
    if ":" in term:
        prefix, local = term.split(":", 1)
        if prefix in prefixes:
            return f"<{prefixes[prefix]}{local}>"
    return term


def _parse_bgp_tokens(tokens: List[str], prefixes: Dict[str, str]) -> List[Triple]:
    """Parse a flat list of tokens into Triple objects (no FILTER/OPTIONAL/UNION)."""
    triples: List[Triple] = []
    i = 0
    last_subject = last_predicate = None
    while i < len(tokens):
        tok = tokens[i]
        low = tok.lower()
        if tok == ".":
            i += 1
            continue
        if low in ("filter", "optional", "union", "bind", "values", "graph"):
            raise SparqlError(f"unexpected keyword in BGP: {tok!r}")
        if tok in {"{", "}"}:
            raise SparqlError(f"unexpected brace in BGP: {tok!r}")
        if tok == ";":
            if last_subject is None:
                raise SparqlError("semicolon without subject")
            i += 1
            if i >= len(tokens) or tokens[i] == ".":
                continue
            subject = last_subject
            predicate = expand_term(tokens[i], prefixes)
            i += 1
        elif tok == ",":
            if last_subject is None or last_predicate is None:
                raise SparqlError("comma without subject/predicate")
            subject = last_subject
            predicate = last_predicate
            i += 1
        else:
            if i + 2 >= len(tokens):
                raise SparqlError(f"incomplete triple pattern near {tok!r}")
            subject = expand_term(tokens[i], prefixes)
            predicate = expand_term(tokens[i + 1], prefixes)
            i += 2
        if i >= len(tokens):
            raise SparqlError("triple pattern missing object")
        obj = expand_term(tokens[i], prefixes)
        i += 1
        triples.append(Triple(subject, predicate, obj))
        last_subject, last_predicate = subject, predicate
    return triples


def parse_bgp_simple(body: str, prefixes: Dict[str, str]) -> List[Triple]:
    """Parse a WHERE sub-body that contains only triple patterns."""
    filtered = []
    for t in tokenize(body):
        if t.upper() in ("ORDER", "LIMIT", "OFFSET"):
            break
        filtered.append(t)
    return _parse_bgp_tokens(filtered, prefixes)


# ---------------------------------------------------------------------------
# Group parser
# ---------------------------------------------------------------------------


def _split_at_top_level_blocks(body: str) -> List[Tuple[str, Optional[str]]]:
    """Split body into (text_before_block, block_content) pairs.

    The last element always has block_content=None (trailing text).
    """
    result: List[Tuple[str, Optional[str]]] = []
    pos = 0
    n = len(body)
    in_iri = in_string = escaped = False
    text_start = 0

    while pos < n:
        ch = body[pos]
        if escaped:
            escaped = False
            pos += 1
            continue
        if ch == "\\" and in_string:
            escaped = True
            pos += 1
            continue
        if ch == "<" and not in_string:
            in_iri = True
        elif ch == ">" and in_iri and not in_string:
            in_iri = False
        elif ch == '"' and not in_iri:
            in_string = not in_string
        elif ch == "{" and not in_iri and not in_string:
            text_before = body[text_start:pos]
            close = matching_brace(body, pos)
            block_content = body[pos + 1 : close]
            result.append((text_before, block_content))
            text_start = close + 1
            pos = close + 1
            continue
        pos += 1

    result.append((body[text_start:], None))
    return result


def _parse_text_segment(
    text: str,
    prefixes: Dict[str, str],
    triples: List[Triple],
    filters: List[str],
) -> None:
    """Parse a text segment (between blocks) for FILTER expressions and triple patterns."""
    # Extract all FILTER(...) blocks using balanced-paren scanning
    remaining = text
    while True:
        m = re.search(r"\bFILTER\s*\(", remaining, re.I)
        if not m:
            break
        paren_start = remaining.index("(", m.start())
        depth = 0
        pos = paren_start
        while pos < len(remaining):
            if remaining[pos] == "(":
                depth += 1
            elif remaining[pos] == ")":
                depth -= 1
                if depth == 0:
                    break
            pos += 1
        filters.append(remaining[paren_start + 1 : pos])
        remaining = remaining[: m.start()] + remaining[pos + 1 :]

    # Strip trailing OPTIONAL keyword (its { } was already extracted as a block)
    remaining = re.sub(r"\bOPTIONAL\s*$", "", remaining.strip(), flags=re.I).strip()

    if remaining:
        triples.extend(parse_bgp_simple(remaining, prefixes))


def parse_group(body: str, prefixes: Dict[str, str]) -> GroupPattern:
    """Parse a WHERE body into a GroupPattern (handles FILTER, OPTIONAL, UNION)."""
    body = strip_comments(body)
    triples: List[Triple] = []
    filters: List[str] = []
    optionals: List[GroupPattern] = []
    unions: List[Tuple[GroupPattern, GroupPattern]] = []

    segments = _split_at_top_level_blocks(body)
    i = 0

    while i < len(segments):
        text, block_content = segments[i]
        text_stripped = text.strip()

        if block_content is not None:
            # Determine block type by what precedes { in the text
            upper_text = text_stripped.upper()

            if upper_text.endswith("OPTIONAL") or re.search(r"\bOPTIONAL\s*$", text_stripped, re.I):
                # OPTIONAL { ... }
                prefix_text = re.sub(r"\bOPTIONAL\s*$", "", text_stripped, flags=re.I).strip()
                if prefix_text:
                    _parse_text_segment(prefix_text, prefixes, triples, filters)
                optionals.append(parse_group(block_content, prefixes))
                i += 1
                continue

            # Check if this is the first arm of a UNION
            if i + 1 < len(segments):
                next_text, next_block = segments[i + 1]
                if next_block is not None and re.match(r"^\s*UNION\s*$", next_text, re.I):
                    arm1 = parse_group(block_content, prefixes)
                    arm2 = parse_group(next_block, prefixes)
                    unions.append((arm1, arm2))
                    # Parse any text before the first arm
                    if text_stripped:
                        _parse_text_segment(text_stripped, prefixes, triples, filters)
                    i += 2
                    continue

            # Standalone sub-group — flatten into current group
            if text_stripped:
                _parse_text_segment(text_stripped, prefixes, triples, filters)
            sub = parse_group(block_content, prefixes)
            triples.extend(sub.triples)
            filters.extend(sub.filters)
            optionals.extend(sub.optionals)
            unions.extend(sub.unions)
            i += 1
            continue

        # Final text-only segment (block_content is None)
        if text_stripped:
            _parse_text_segment(text_stripped, prefixes, triples, filters)
        i += 1

    return GroupPattern(triples=triples, filters=filters, optionals=optionals, unions=unions)


# ---------------------------------------------------------------------------
# ORDER BY parser
# ---------------------------------------------------------------------------


def _parse_order_by(text: str) -> Tuple[Tuple[str, bool], ...]:
    m = ORDERBY_RE.search(text)
    if not m:
        return ()
    order_text = m.group(1).strip()
    result: List[Tuple[str, bool]] = []
    # DESC(?var)
    for var in re.findall(r"DESC\s*\(\s*\?(\w+)\s*\)", order_text, re.I):
        result.append((var, False))
    # ASC(?var) or bare ?var (treated as ascending)
    remainder = re.sub(r"(?:ASC|DESC)\s*\([^)]+\)", "", order_text, flags=re.I)
    for var in re.findall(r"\?(\w+)", remainder):
        result.append((var, True))
    return tuple(result)


# ---------------------------------------------------------------------------
# Query parser
# ---------------------------------------------------------------------------


def unique(values: Iterable[str]) -> List[str]:
    seen: set = set()
    out: List[str] = []
    for value in values:
        if value not in seen:
            out.append(value)
            seen.add(value)
    return out


def parse_query(text: str) -> Query:
    prefixes = {m.group(1): m.group(2) for m in PREFIX_RE.finditer(text)}
    select_match = SELECT_RE.search(text)
    if not select_match:
        raise SparqlError("unsupported query: missing SELECT ... WHERE")
    select_clause = select_match.group(2).strip()
    select = [] if select_clause == "*" else unique(VAR_RE.findall(select_clause))
    limit_match = LIMIT_RE.search(text)
    limit = int(limit_match.group(1)) if limit_match else None
    group = parse_group(where_body(text), prefixes)
    order_by = _parse_order_by(text)
    return Query(
        prefixes=prefixes,
        select=select,
        group=group,
        distinct=bool(select_match.group(1)),
        limit=limit,
        order_by=order_by,
    )


def parse_endpoints(text: str) -> List[Endpoint]:
    endpoints = []
    for idx, match in enumerate(ENDPOINT_RE.finditer(text)):
        raw_id = re.sub(r"[^A-Za-z0-9_.-]+", "_", match.group(1)).strip("_")
        endpoints.append(Endpoint(raw_id or f"endpoint_{idx}", match.group(2)))
    if not endpoints:
        raise SparqlError("no sd:endpoint entries found in config")
    return endpoints



__all__ = ["expand_term", "parse_bgp_simple", "parse_endpoints", "parse_group", "parse_query", "strip_comments", "tokenize", "tokenize_filter", "unique", "VAR_RE", "where_body"]
